S2VX.Game/SongSelection/SongSelectionScreen.cs

```diff
@@ -1,37 +1,150 @@
 ﻿using osu.Framework.Allocation;
+using osu.Framework.Audio;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
-using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Input.Events;
+using osu.Framework.IO.Stores;
+using osu.Framework.Platform;
 using osu.Framework.Screens;
-using osuTK;
-using S2VX.Game.Editor;
+using osuTK.Graphics;
+using osuTK.Input;
+using S2VX.Game.SongSelection.Containers;
+using S2VX.Game.SongSelection.UserInterface;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 
 namespace S2VX.Game.Play {
     public class SongSelectionScreen : Screen {
-        private static Vector2 BtnSize { get; } = new Vector2(100, 30);
+        [Resolved]
+        private ScreenStack Screens { get; set; }
+        [Resolved]
+        private AudioManager Audio { get; set; }
 
-        private Button BtnEdit { get; } = new BasicButton() {
-            Size = BtnSize,
-            Text = "Edit"
-        };
+        public string CurSelectionPath { get; set; } = "Stories";
 
-        private Button BtnPlay { get; } = new BasicButton() {
-            Size = BtnSize,
-            Text = "Play"
-        };
+        private TextFlowContainer CurSelectionPathTxt { get; set; }
+        private NativeStorage Storage { get; set; }
+        private StorageBackedResourceStore CurLevelResourceStore { get; set; }
+
+        private List<Drawable> CreateSelectionItems() {
+            var selectionItems = new List<Drawable>();
+            var dirs = Storage.GetDirectories("");
+            foreach (var dir in dirs) {
+                selectionItems.Add(new SelectedItemDisplay {
+                    ItemName = dir,
+                    SongSelectionScreen = this,
+                });
+            }
+            return selectionItems;
+        }
+
+        private (bool, string, string) DirectoryContainsStory(string dir) {
+            var story = Storage.GetFiles(dir, "*.s2ry");
+            var song = Storage.GetFiles(dir, "audio.mp3");
+            return (story.Count() == 1 && song.Count() == 1, story.FirstOrDefault(), song.FirstOrDefault());
+        }
+
+        private bool DirectoryContainsDirectories(string dir) => Storage.GetDirectories(dir).Any();
+
+        // Go up one level by exiting and thus popping ourself out from the ScreenStack
+        public void DoExit() {
```

Comment:
Can you just override IScreen's OnExiting instead? I think this will also remove the need to pass SongSelectionScreen everywhere.

Reply:
If I do that then I'll need to move the logic for checking for root level to BorderOuterBox's OnClick. Is that preferred?

Reply:
The OnExit function says that it will prevent exit if you return true, can't you use this? 

Isn't the current selection screen always on top? Can you not do some kind of generic screen poptop to remove the top screen?

Reply:
Thanks for pointing out the return true behavior. Will do.